File: ios/RocChat/Calls/MuLaw.swift
/**
 * RocChat iOS — G.711 μ-law codec
 *
 * Lossy 2:1 audio compression (PCM16 → 8-bit μ-law).
 * ITU-T G.711 standard, public domain since 1972.
 * Zero third-party dependencies. ~10 µs per 20 ms frame on A17.
 */

import Foundation

enum MuLaw {
    private static let bias: Int32 = 0x84
    private static let clip: Int32 = 32_635

    static func encode(pcm16: Data) -> Data {
        let count = pcm16.count / MemoryLayout<Int16>.size
        var out = Data(count: count)
        pcm16.withUnsafeBytes { raw in
            guard let src = raw.baseAddress?.assumingMemoryBound(to: Int16.self) else { return }
            out.withUnsafeMutableBytes { outRaw in
                guard let dst = outRaw.baseAddress?.assumingMemoryBound(to: UInt8.self) else { return }
                for i in 0..<count {
                    dst[i] = encodeSample(src[i])
                }
            }
        }
        return out
    }

    static func decode(mulaw: Data) -> Data {
        var out = Data(count: mulaw.count * MemoryLayout<Int16>.size)
        mulaw.withUnsafeBytes { raw in
            guard let src = raw.baseAddress?.assumingMemoryBound(to: UInt8.self) else { return }
            out.withUnsafeMutableBytes { outRaw in
                guard let dst = outRaw.baseAddress?.assumingMemoryBound(to: Int16.self) else { return }
                for i in 0..<mulaw.count {
                    dst[i] = decodeSample(src[i])
                }
            }
        }
        return out
    }

    @inline(__always)
    private static func encodeSample(_ sample: Int16) -> UInt8 {
        var s = Int32(sample)
        let sign: Int32 = (s < 0) ? 0x7F : 0xFF
        if s < 0 { s = -s }
        if s > clip { s = clip }
        s += bias
        var exponent: Int32 = 7
        var expMask: Int32 = 0x4000
        while (s & expMask) == 0 && exponent > 0 {
            exponent -= 1
            expMask >>= 1
        }
        let mantissa = (s >> (exponent == 0 ? 4 : (exponent + 3))) & 0x0F
        let byte = ~(Int32((exponent << 4) | mantissa)) & sign
        return UInt8(truncatingIfNeeded: byte)
    }

    @inline(__always)
    private static func decodeSample(_ byte: UInt8) -> Int16 {
        let b = Int32(~byte)
        let sign = b & 0x80
        let exponent = (b >> 4) & 0x07
        let mantissa = b & 0x0F
        var sample = ((mantissa << 3) + bias) << exponent
        sample -= bias
        if sign != 0 { sample = -sample }
        // Clamp to Int16 range
        if sample > Int32(Int16.max) { sample = Int32(Int16.max) }
        if sample < Int32(Int16.min) { sample = Int32(Int16.min) }
        return Int16(sample)
    }
}

File: android/app/src/main/kotlin/com/rocchat/calls/MuLaw.kt
/**
 * RocChat Android — G.711 μ-law codec
 *
 * Mirrors MuLaw.swift — same encode/decode algorithm, byte-compatible on the wire.
 * Zero third-party dependencies.
 */

package com.rocchat.calls

object MuLaw {
    private const val BIAS = 0x84
    private const val CLIP = 32_635

    fun encode(pcm16: ByteArray): ByteArray {
        val count = pcm16.size / 2
        val out = ByteArray(count)
        for (i in 0 until count) {
            val lo = pcm16[i * 2].toInt() and 0xFF
            val hi = pcm16[i * 2 + 1].toInt()
            val sample = ((hi shl 8) or lo).toShort().toInt()
            out[i] = encodeSample(sample).toByte()
        }
        return out
    }

    fun decode(mulaw: ByteArray): ByteArray {
        val out = ByteArray(mulaw.size * 2)
        for (i in mulaw.indices) {
            val sample = decodeSample(mulaw[i].toInt() and 0xFF)
            out[i * 2]     = (sample and 0xFF).toByte()
            out[i * 2 + 1] = ((sample ushr 8) and 0xFF).toByte()
        }
        return out
    }

    private fun encodeSample(sample: Int): Int {
        var s = sample
        val sign = if (s < 0) 0x7F else 0xFF
        if (s < 0) s = -s
        if (s > CLIP) s = CLIP
        s += BIAS
        var exponent = 7
        var expMask = 0x4000
        while ((s and expMask) == 0 && exponent > 0) {
            exponent--
            expMask = expMask ushr 1
        }
        val shift = exponent + 3
        val mantissa = (s ushr shift) and 0x0F
        return (((exponent shl 4) or mantissa).inv()) and sign and 0xFF
    }

    private fun decodeSample(byte: Int): Int {
        val b = byte.inv() and 0xFF
        val sign = b and 0x80
        val exponent = (b ushr 4) and 0x07
        val mantissa = b and 0x0F
        var sample = ((mantissa shl 3) + BIAS) shl exponent
        sample -= BIAS
        if (sign != 0) sample = -sample
        if (sample > Short.MAX_VALUE.toInt()) sample = Short.MAX_VALUE.toInt()
        if (sample < Short.MIN_VALUE.toInt()) sample = Short.MIN_VALUE.toInt()
        return sample
    }
}
